File: speed-run/src/main/kotlin/com/ryangwaite/protocol/Messages.kt
package com.ryangwaite.protocol

import com.ryangwaite.models.HostQuestionSummary
import com.ryangwaite.models.LeaderboardItem
import com.ryangwaite.models.ParticipantQuestionSummary
import com.ryangwaite.models.QuestionAndAnswers
import kotlinx.serialization.Serializable

sealed class ProtocolMsg {
    @Serializable
    enum class Type {
        // Messages from the quiz host to the server which a response is not sent back for
        `HOST-CONFIG`,
        `HOST-START`,
        // Messsages sent from the server to the quiz host which a response is not sent back for
        `NOTIFY-HOST-QUIZ-SUMMARY`,
        // Messages sent from the server to a single quiz participant which a response is not sent back for
        `NOTIFY-PARTICIPANT-QUIZ-SUMMARY`,
        // Request and response messages between quiz host and server
        `REQUEST-HOST-QUESTIONS`,
        `RESPONSE-HOST-QUESTIONS`,
        // Messages from a participant to the server which a response is not sent back for
        `PARTICIPANT-CONFIG`,
        `PARTICIPANT-ANSWER`,
        `PARTICIPANT-ANSWER-TIMEOUT`,
        // Request and response messages between a quiz participant and server
        `REQUEST-PARTICIPANT-QUESTION`,
        `RESPONSE-PARTICIPANT-QUESTION`,
        // Messages which are sent from the server to the host and all participants which a response is not sent back for
        `BROADCAST-PARTICIPANT-CONFIG`,
        `BROADCAST-START`,
        `LEADERBOARD`,
        `BROADCAST-PARTICIPANT-FINISHED`,
        `BROADCAST-QUIZ-FINISHED`,
    }
}

@Serializable
data class HostConfigMsg(
    val quizName: String,
    val categories: List<String>,
    val duration: Int,
    val selectedQuestionIndexes: List<Int>,
): ProtocolMsg()

@Serializable
class HostStartMsg: ProtocolMsg()

@Serializable
class RequestHostQuestionsMsg: ProtocolMsg()

@Serializable
data class ResponseHostQuestionsMsg(
    val questions: List<QuestionAndAnswers>,
): ProtocolMsg()

@Serializable
data class NotifyHostQuizSummaryMsg(
    val totalTimeElapsed: Long, // seconds
    val avgAnswerTime: Int,  // milliseconds
    val questions: List<HostQuestionSummary>,
): ProtocolMsg()

@Serializable
data class NotifyParticipantQuizSummaryMsg(
    // These metrics are just for this participant
    val totalTimeElapsed: Long, // seconds
    val avgAnswerTime: Int, // milliseconds
    val questions: List<ParticipantQuestionSummary>,
): ProtocolMsg()

@Serializable
data class ParticipantConfigMsg(
    val name: String,
): ProtocolMsg()

@Serializable
data class ParticipantAnswerMsg(
    val questionIndex: Int,
    val selectedOptionIndexes: List<Int>,
    val answerDuration: Int, // milliseconds
): ProtocolMsg()

@Serializable
data class ParticipantAnswerTimeoutMsg(
    val questionIndex: Int
): ProtocolMsg()

@Serializable
data class RequestParticipantQuestionMsg(
    val questionIndex: Int,
): ProtocolMsg()

@Serializable
data class ResponseParticipantQuestionMsg(
    val questionIndex: Int,
    val question: String,
    val options: List<String>,
    val numberOfOptionsToSelect: Int,
): ProtocolMsg()

@Serializable
data class BroadcastParticipantConfigMsg(
    val userId: String,
    val name: String,
): ProtocolMsg()

@Serializable
data class BroadcastStartMsg(
    val startTimeEpochSecs: Long,
    val questionDuration: Int,
    val numberOfQuestions: Int,
): ProtocolMsg()

@Serializable
data class LeaderboardMsg(
    val leaderboard: List<LeaderboardItem>
): ProtocolMsg()

@Serializable
data class BroadcastParticipantFinishedMsg(
    val totalFinishedParticipants: Int,
): ProtocolMsg()

@Serializable
object BroadcastQuizFinishedMsg : ProtocolMsg()

File: ui/src/api/protocol/messages.ts
import { IHostQuestion, IHostQuestionSummary, ILeaderboardItem, IParticipantQuestionSummary, IQuestionAndAnswers } from "../../types"

/********************* OUTBOUND *********************/
export const PARTICIPANT_CONFIG = "PARTICIPANT-CONFIG"
export function participantConfigMsg(name: string) {
    return {
        name
    }
}
export type ParticipantConfigMsgType = ReturnType<typeof participantConfigMsg>

export const REQUEST_HOST_QUESTIONS = "REQUEST-HOST-QUESTIONS"
export function requestHostQuestions() {
    return {}
}
export type RequestHostQuestionsMsgType = ReturnType<typeof requestHostQuestions>

export const REQUEST_HOST_QUIZ_SUMMARY = "REQUEST-HOST-QUIZ-SUMMARY"
export function requestHostQuizSummary() {
    return {}
}
export type RequestHostQuizSummaryMsgType = ReturnType<typeof requestHostQuizSummary>

export const HOST_CONFIG = "HOST-CONFIG"
export function hostConfig(quizName: string, categories: string[], duration: number, selectedQuestionIndexes: number[]) {
    return {
        quizName,
        categories,
        duration,
        selectedQuestionIndexes,
    }
}
export type HostConfigMsgType = ReturnType<typeof hostConfig>

export const REQUEST_PARTICIPANT_QUESTION = "REQUEST-PARTICIPANT-QUESTION"
export function requestParticipantQuestion(questionIndex: number) {
    return {
        questionIndex
    }
}
export type RequestParticipantQuestionMsgType = ReturnType<typeof requestParticipantQuestion>

export const PARTICIPANT_ANSWER = "PARTICIPANT-ANSWER"
export function participantAnswer(questionIndex: number, selectedOptionIndexes: number[], answerDuration: number) {
    return {
        questionIndex,
        selectedOptionIndexes,
        answerDuration
    }
}
export type ParticipantAnswerMsgType = ReturnType<typeof participantAnswer>

export const PARTICIPANT_ANSWER_TIMEOUT = "PARTICIPANT-ANSWER-TIMEOUT"
export function participantAnswerTimeout(questionIndex: number) {
    return {
        questionIndex
    }
}
export type ParticipantAnswerTimeoutMsgType = ReturnType<typeof participantAnswerTimeout>

/********************* INBOUND *********************/
export const BROADCAST_LEADERBOARD = "BROADCAST-LEADERBOARD"
export interface BroadcastLeaderboardMsgType {
    leaderboard: ILeaderboardItem[]
}

export const RESPONSE_HOST_QUESTIONS = "RESPONSE-HOST-QUESTIONS"
export interface ResponseHostQuestionsMsgType {
    questions: IQuestionAndAnswers[]
}

export const RESPONSE_HOST_QUIZ_SUMMARY = "RESPONSE-HOST-QUIZ-SUMMARY"
export interface ResponseHostQuizSummaryMsgType {
    totalTimeElapsed: number, // milliseconds
    questions: IHostQuestion[],
}

export const BROADCAST_START = "BROADCAST-START"
export interface BroadcastStartMsgType {
    startTimeEpochSecs: number,
    questionDuration: number,
    numberOfQuestions: number,
}

export const RESPONSE_PARTICIPANT_QUESTION = "RESPONSE-PARTICIPANT-QUESTION"
export interface ResponseParticipantQuestionMsgType {
    questionIndex: number,
    question: string,
    options: string[],
    numberOfOptionsToSelect: number,
}

export const NOTIFY_HOST_QUIZ_SUMMARY = "NOTIFY-HOST-QUIZ-SUMMARY"
export interface NotifyHostQuizSummaryMsgType {
    totalTimeElapsed: number,   // seconds
    avgAnswerTime: number,      // milliseconds
    questions: IHostQuestionSummary[],
}

export const NOTIFY_PARTICIPANT_QUIZ_SUMMARY = "NOTIFY-PARTICIPANT-QUIZ-SUMMARY"
export interface NotifyParticipantQuizSummaryMsgType {
    totalTimeElapsed: number,  //milliseconds
    avgAnswerTime: number,  // milliseconds
    questions: IParticipantQuestionSummary[],
}

export const BROADCAST_QUIZ_FINISHED = "BROADCAST-QUIZ-FINISHED"
export interface BroadcastQuizFinishedMsgType {}